WalletWasabi/Wallets/Wallet.cs

```diff
@@ -513,13 +508,76 @@ private async Task LoadDummyMempoolAsync()
 		}
 	}
 
-	private async Task ProcessFilterModelAsync(FilterModel filterModel, CancellationToken cancel)
+	/// <summary>
+	/// Go through the filters and queue to download the matches.
+	/// </summary>
+	public async Task PerformWalletSynchronizationAsync(SyncType syncType, CancellationToken cancel)
+	{
+		var startingHeight = syncType == SyncType.Turbo ? 
+			new Height(KeyManager.GetBestTurboSyncHeight() + 1) : 
+			new Height(KeyManager.GetBestHeight() + 1);
+
+		await BitcoinStore.IndexStore.ForeachFiltersAsync(
+			async (filterModel) => await ProcessFilterModelAsync(filterModel, syncType, cancel).ConfigureAwait(false),
+			startingHeight,
+			cancel).ConfigureAwait(false);
+	}
+
+	/// <summary>
+	/// Return the keys to test against the filter depending on the height of the filter and the type of synchronization.
+	/// </summary>
+	/// <param name="filterHeight"></param>
+	/// <param name="syncType">First sync of TurboSync, second one, or complete synchronization</param>
+	/// <returns>Keys to test against this filter</returns>
+	/// <seealso href="https://github.com/zkSNACKs/WalletWasabi/issues/10219">TurboSync specification.</seealso>
+	private List<byte[]> GetScriptPubKeysToTest(Height filterHeight, SyncType syncType)
+	{
+		if (syncType == SyncType.Complete)
+		{
+			return KeyManager.GetHdPubKeysWithScriptBytes().Select(x => x.ScriptBytes).ToList();
+		}
+
+		var allKeysWithScriptBytes = KeyManager.GetHdPubKeysWithScriptBytes();
+		
+		IEnumerable<byte[]> keysToTest;
+		if (syncType == SyncType.Turbo)
+		{
+			// First sync during TurboSync: test only external keys or internal that never received or still had coins at filter height.
+			keysToTest = allKeysWithScriptBytes
+				.Where(
+					x =>
+						x.HdPubKey.LatestSpendingHeight is null ||
+						(Height)x.HdPubKey.LatestSpendingHeight >= filterHeight)
+				.Select(x => x.ScriptBytes);
+		}
+		else
+		{
+			// Second sync during TurboSync: test all other keys (internal keys that received and already spent their coins at filter height).
+			keysToTest = allKeysWithScriptBytes.Where(
+				x =>
+					x.HdPubKey.LatestSpendingHeight is not null &&
+					(Height)x.HdPubKey.LatestSpendingHeight < filterHeight)
+				.Select(x => x.ScriptBytes);
+		}
```

Comment:
I would suggest you to move the `if` statement somewhere else to avoid having this almost exact duplicated code. Here an alternative for you to consider:

```suggestion
			Func<HdPubKey, bool> stepPredicate = syncType == SyncType.Turbo
				? hdPubKey => hdPubKey.LatestSpendingHeight is null || (Height)hdPubKey.LatestSpendingHeight >= filterHeight
				: hdPubKey => hdPubKey.LatestSpendingHeight is not null || (Height)hdPubKey.LatestSpendingHeight! < filterHeight;

			IEnumerable<HdPubKey> keysToTest = KeysCache.Where(stepPredicate);
```